compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp

```diff
@@ -507,4 +512,266 @@ LogicalResult splitLogicalObjectFifos(
+    SmallVector<OpFoldResult> sourceSizes = consumer.getSourceMixedSizes();
+    SmallVector<OpFoldResult> sourceStrides = consumer.getSourceMixedStrides();
+
+    std::optional<int64_t> sourceSize =
+        getConstantIntValue(sourceSizes[splitDim]);
+    std::optional<int64_t> sourceOffset =
+        getConstantIntValue(sourceOffsets[splitDim]);
+    if (!sourceSize || !sourceOffset) {
+      return consumer.emitOpError()
+             << "expected a static source offset and size on index: "
+             << splitDim;
+    }
+    if (sourceSize.value() != 1) {
+      return consumer.emitOpError() << "only a static size of 1 is currently "
+                                       "supported on the split index";
+    }
+    assert(sourceOffset < newObjFifos.size() &&
+           "the sourceOffset should be smaller than the number of objectFifos");
+
+    // The index has been encoded as source offset for the split dimension, so
+    // fetch the corresponding objectFifo based on it.
+    AMDAIE::LogicalObjectFifoFromMemrefOp newObjFifo =
+        newObjFifos[sourceOffset.value()];
+    sourceOffsets[splitDim] = rewriter.getIndexAttr(0);
+    rewriter.setInsertionPoint(consumer);
+    auto newDmaOp = rewriter.create<AMDAIE::DmaCpyNdOp>(
+        consumer.getLoc(), consumer.getTarget(),
+        consumer.getTargetMixedOffsets(), consumer.getTargetMixedSizes(),
+        consumer.getTargetMixedStrides(), newObjFifo, sourceOffsets,
+        sourceSizes, sourceStrides);
+    rewriter.replaceOp(consumer, newDmaOp);
+  }
+
+  return success();
+}
+
+/// Split DmaCpyNd ops between L2 and L3 memory spaces.
+LogicalResult splitDoublyStridedOp(IRRewriter &rewriter,
+                                   AMDAIE::DmaCpyNdOp op) {
+  // Get the split dim from L2 side objectfifo.
+  LogicalObjectFifoFromMemrefOp srcObjectFifo = op.getSourceObjectFifo();
+  LogicalObjectFifoFromMemrefOp tgtObjectFifo = op.getTargetObjectFifo();
+  FailureOr<size_t> maybeSplitDim;
+  ArrayRef<int64_t> memrefShape;
+  bool isL2Target;
+  if (srcObjectFifo.getMemorySpaceAsUInt() == 1) {
+    isL2Target = false;
+    maybeSplitDim = getSplitDim(srcObjectFifo);
+    memrefShape = srcObjectFifo.getMemrefType().getShape();
+  } else if (tgtObjectFifo.getMemorySpaceAsUInt() == 1) {
+    isL2Target = true;
+    maybeSplitDim = getSplitDim(tgtObjectFifo);
+    memrefShape = tgtObjectFifo.getMemrefType().getShape();
+  } else {
+    return op.emitOpError()
+           << "the input dma should have source or target in L2 memory space";
+  }
+  if (failed(maybeSplitDim)) return failure();
+  size_t splitDim = maybeSplitDim.value();
+
+  // Both outer dims are 1, no need to split, return success.
+  if (splitDim == memrefShape.size()) return success();
+  int64_t splitFactor = memrefShape[splitDim];
+  if (splitFactor == 1) return success();
+
+  FailureOr<bool> l2DmaTransposed = isL2DmaTransposed(op, isL2Target);
+  if (failed(l2DmaTransposed)) return failure();
```

Comment:
Why is this needed?

Reply:
It's because the shape of objectFifo and the target size of DmaCpNdOp can be different if we select transpose on target option within `ConvertToDma` pass. The splitDim we get earlier in this function is only the split dim in objectFifo, but not the split dim for the target size in DmaCpNdOp.

For example, in this example `%3 = amdaie.dma_cpy_nd(%0[0, 0, 0, 0] [1, 32, 2, 32] [2048, 32, 1024, 1], %1[0, %2] [32, 64] [128, 1]) : (!amdaie.logicalobjectfifo<memref<1x2x32x32xi32, 1 : i32>>, !amdaie.logicalobjectfifo<memref<128x128xi32>>)`,

the splitDim for objectfifo is index 1 while the splitDimTarget should be index 2.